### ISSUE TITLE
[Bug] Age Restricted Videos do not play

### ISSUE THREAD

**Issue by SimoonDude** [iv-org/invidious#2189]
Videos that have been marked age-restricted don't play
Any video with age-restriction doesn't work. All you can see is the thumbnail. I've tested this on all public instances.

1. Go to any age-restricted video
2. See error

Even the play button doesn't work
![image](https://user-images.githubusercontent.com/86322479/122985905-65738200-d3a7-11eb-9a28-42e6dbcd8296.png)

**Additional context**
 - Browser: Latest Windows version of Chrome
 - OS: Win10 x64

**Comment by Mennaruuk** [comment]
Yeah, I’ve been facing the same issue. It seems like a recent problem. I don’t know if it’s related to what is going on at [youtube-dl](https://github.com/ytdl-org/youtube-dl/issues/28578). But I couldn’t download using youtube-dl either, and I’m not in the E.U. nor do I have a Google account.

**Comment by unixfox** [comment]
Can someone post a couple of age-restricted videos here?

**Comment by Mennaruuk** [comment]
`https://www.youtube.com/watch?v=5gZCjMjoilo`

**Comment by TheFrenchGhosty** [comment]
A 20$ bounty has been added to this issue.

Anyone opening a PR fixing this issue, will receive 20$ (in BTC) from the Invidious project.

More details: https://github.com/iv-org/invidious/issues/1898

**Comment by TheFrenchGhosty** [comment]
Note: the youtube-dl/yt-dlp people found a way to get access to the video by (I think) using the mobile website. However this only works with videos that can be embedded. Having this solution implemented is "enough" for you to get the bounty.

Related: https://github.com/yt-dlp/yt-dlp/commit/4e6767b5f2e2523ebd3dd1240584ead53e8c8905

**Comment by pukkandan** [comment]
Just add `&c=TVHTML5&cver=6.20180913` to https://github.com/iv-org/invidious/blob/993d731c92f6f07e4bee4a222c02e1e47d7f9d6e/src/invidious/videos.cr#L994

Note that this is a temporary fix and we expect it to break in the near future. See https://github.com/yt-dlp/yt-dlp/pull/328 for some fallbacks

**Comment by syeopite** [comment]
The `x-youtube-client-version` header would also have to be set to `6.20180913`

**Comment by pukkandan** [comment]
> The `x-youtube-client-version` header would also have to be set to `6.20180913`

No. `get_video_info` doesnt need `x-youtube-client-version` header

**Comment by syeopite** [comment]
Yep. However, Invidious appends those headers by default as seen in here:
https://github.com/iv-org/invidious/blob/993d731c92f6f07e4bee4a222c02e1e47d7f9d6e/src/invidious/helpers/utils.cr#L4-L17

 In order to get this fix to work, `x-youtube-client-version` would also have to be set to `6.20180913`

**Comment by pukkandan** [comment]
hm.. that is a problem. The header for all other requests should still be `2.20210407.08.00`

**Comment by syeopite** [comment]
Nope. If we want to use the connection pool, all of those headers would get added no matter what. The only workaround for this is to set `x-youtube-client-version` to `6.20180913` as well since the `add_yt_headers` function doesn't override existing headers.

**Comment by neurodiverseEsoteric** [comment]
Can this userscript be pasted somewhere into invidious? It worked for https://www.youtube.com/watch?v=KgmoMO66uPg : 

 https://raw.githubusercontent.com/zerodytrash/Simple-YouTube-Age-Restriction-Bypass/main/Simple-YouTube-Age-Restriction-Bypass.user.js#bypass=true

**Comment by unixfox** [comment]
> Can this userscript be pasted somewhere into invidious? It worked for [youtube.com/watch?v=KgmoMO66uPg](https://www.youtube.com/watch?v=KgmoMO66uPg) :
> 
> [zerodytrash/Simple-YouTube-Age-Restriction-Bypass@`main`/Simple-YouTube-Age-Restriction-Bypass.user.js#bypass=true (raw)](https://raw.githubusercontent.com/zerodytrash/Simple-YouTube-Age-Restriction-Bypass/main/Simple-YouTube-Age-Restriction-Bypass.user.js#bypass=true)

That's already what the PR #2205 is doing.

**Comment by neurodiverseEsoteric** [comment]
Oh ok, nevermind then....

**Comment by 89z** [comment]
https://github.com/iv-org/invidious/issues/2189
https://github.com/yt-dlp/yt-dlp/issues/319
https://github.com/ytdl-org/youtube-dl/issues/29086

**Comment by unixfox** [comment]
Reopening as not every age restricted videos are unblocked.

**Comment by 89z** [comment]
@unixfox can you name a problem video? I think I have a fix

**Comment by unixfox** [comment]
> @unixfox can you name a problem video? I think I have a fix

Well this video used to not work with the fix for bypassing age restricted videos: https://www.youtube.com/watch?v=cIVPYWyPqKA but now it works :thinking:. I can't find a video that we can't bypass though.

We are still interested in your potential fix though.

**Comment by 89z** [comment]
Id rather not present a solution, unless a problem exists. If you find another problem video let me know

**Comment by syeopite** [comment]
> https://github.com/yt-dlp/yt-dlp/pull/328#issuecomment-864117168

`Cr381pDsSsA`  is one that still doesn't play even after #2205

**Comment by 89z** [comment]
> `Cr381pDsSsA` is one that still doesn't play

So this is one that gives you this on the normal page [1]:

~~~
Sign in to confirm your age
~~~

and gives you this on the embed page [2]:

~~~
This video is age-restricted and only available on YouTube
~~~

To fix, you can either make a request to `/youtubei/v1/player` with cookies [3], or you can make a request to `/get_video_info` using the `ANDROID` client [4].

1. https://www.youtube.com/watch?v=Cr381pDsSsA
2. https://www.youtube.com/embed/Cr381pDsSsA
3. https://github.com/kkdai/youtube/issues/202#issuecomment-877798231
4. https://github.com/kkdai/youtube/issues/202#issuecomment-878454739

**Comment by pukkandan** [comment]
Using cookies obviously work, the challenge is making it work without logging in

> you can make a request to /get_video_info using the ANDROID client [4].

Doesn't work 
https://www.youtube.com/get_video_info?video_id=Cr381pDsSsA&eurl=https%3A%2F%2Fyoutube.googleapis.com%2Fv%2FCr381pDsSsA&html5=1&c=ANDROID&cver=16.20

**Comment by 89z** [comment]
Huh weird. If I do the normal request I get this:

~~~
https://www.youtube.com/get_video_info?c=ANDROID&cver=16.05&eurl=https%3A%2F%2Fwww.youtube.com&html5=1&video_id=Cr381pDsSsA
Playback on other websites has been disabled by the video owner.
~~~

I had found that adding `el` sometime help, but not this time:

~~~
https://www.youtube.com/get_video_info?c=ANDROID&cver=16.05&el=detailpage&eurl=https%3A%2F%2Fwww.youtube.com&html5=1&video_id=Cr381pDsSsA
This video may be inappropriate for some users.
~~~

maybe a different `el` value? I have also seen `el=embedded`, maybe others exist too?

**Comment by 89z** [comment]
After much research, it seems some video simply must use authentication. However I disagree with the Cookie approach commonly implemented. I think an OAuth approach is a better experience. I have some information here [1] if any interested, I am working on an implementation in Go.

1. https://github.com/89z/mech/tree/28e2111d/research/2021-07-21

**Comment by SunburntRock89** [comment]
I have a strange feeling that all of get_video_info has been removed as I've seen it 404 on everything I've thrown it - age restricted or not. Some googling shows that a while back they added more required fields which could be a possiblity, but I'd say they probably just entirely removed this endpoint to stop people from bypassing age restriction for legal reasons.

**Comment by pukkandan** [comment]
> I have a strange feeling that all of get_video_info has been removed as I've seen it 404 on everything I've thrown it

Seems like it. We also couldn't find a set of parameters that works. the embed-workaround still works for embeddable videos, but auth seems necessary for everything else now

**Comment by D3V1LC0D3R** [comment]
there seems to be a greasyfork script to fix this issue, so does this code help?
https://greasyfork.org/en/scripts/375525-youtube-age-verification-bypass/code

**Comment by pukkandan** [comment]
> there seems to be a greasyfork script to fix this issue, so does this code help?
> https://greasyfork.org/en/scripts/375525-youtube-age-verification-bypass/code

This is what https://github.com/iv-org/invidious/pull/2205 did. That method no longer works

**Comment by reallyuniquename** [comment]
> the embed-workaround still works for embeddable videos

Could you point out which one is that?

**Comment by reallyuniquename** [comment]
> the embed-workaround still works for embeddable videos

@pukkandan Could you point out which one is that?

**Comment by pukkandan** [comment]
You need to re-request the `/browse` API with the `WEB_EMBEDDED` client instead of `WEB` (other clients like ANDROID_EMBEDDED, IOS_EMBEDDED also work). See related params [here](https://github.com/yt-dlp/yt-dlp/blob/8242bf220deac405f164675fdff16e8f6fd044fc/yt_dlp/extractor/youtube.py#L344-L357)

This method can bypass embeddable videos like `HtVdAasjOgU`, but others like `Tq92D6wQ1mg` will still not work

**Comment by pukkandan** [comment]
There's a new age-gate method: https://github.com/yt-dlp/yt-dlp/issues/574#issuecomment-887171136

**Comment by unixfox** [comment]
Updated method created in PR https://github.com/iv-org/invidious/pull/2329, feel free to test and submit your feedback.

**Comment by unixfox** [comment]
Reopening like https://github.com/iv-org/invidious/issues/2189#issuecomment-880080346

**Comment by sriegler** [comment]
I just stumbled across this issue. It worked in my build from 2022.01.21 and failed in the latest. After some testing, according to git disect commit 99091e919c9af56c27ca8aebd790c3b64b564f78 broke it again:

last good: 2022.02.03-6ddbccbc
first bad: 2022.02.03-99091e91

**Comment by unixfox** [comment]
Message from @TheFrenchGhosty:

> The age restriction bypass isn't working consistently, it will only properly work after ~5 requests (just refreshing the page on the instance works). When it works once, it will work consistently everytime (because of caching I'm guessing).
> 
> **Screenshots**
> 
> ![error](https://user-images.githubusercontent.com/47571719/154084411-a6e3e59a-864e-4d89-98ce-a323c46380a9.png)
> 
> **Additional context**
> 
> I can provide some examples videos IDs at request.
> 
> On a related note: the age restricted videos, don't have descriptions.

**Comment by pukkandan** [comment]
Youtube is rolling out some changes that breaks [tier 3](https://github.com/yt-dlp/yt-dlp/pull/575#issuecomment-888837000) age-gate bypass. There is currently no known workaround (other than [passing the request through a proxy](https://github.com/zerodytrash/Simple-YouTube-Age-Restriction-Bypass/tree/main/account-proxy)). yt-dlp and newpipe are also affected

**Comment by girst** [comment]
@pukkandan: is there a yt-dlp or newpipe issue i can follow for this?

**Comment by unixfox** [comment]
New method for bypass: https://github.com/yt-dlp/yt-dlp/pull/3233

**Comment by unixfox** [comment]
New method submitted in https://github.com/iv-org/invidious/pull/2996

**Comment by TheFrenchGhosty** [comment]
Here's 4 SFW videos that are age restricted, for testing purpose:

https://youtube.com/watch?v=FISUqoODw9Q
https://youtube.com/watch?v=KVjHwrKJg8k
https://youtube.com/watch?v=h16HA4XjyD8
https://youtube.com/watch?v=FISUqoODw9Q

**Comment by girst** [comment]
On Tue, Mar 29, 2022 at 01:00:24PM -0700, TheFrenchGhosty wrote:
>Here's 4 SFW videos that are age restricted, for testing purpose:

there's a 4-tiered list of videos that's used to check agegate bypasses:
https://github.com/yt-dlp/yt-dlp/pull/575#issuecomment-888837000

**Comment by unixfox** [comment]
> On Tue, Mar 29, 2022 at 01:00:24PM -0700, TheFrenchGhosty wrote:
> >Here's 4 SFW videos that are age restricted, for testing purpose:
> 
> there's a 4-tiered list of videos that's used to check agegate bypasses:
> https://github.com/yt-dlp/yt-dlp/pull/575#issuecomment-888837000
> 

Tiers 1,2 and 3 seems to work with the new method, which is great.

**Comment by tuxayo** [comment]
Should this ticket be closed now that #2996 has been merged? Or is there still and issue?

**Comment by unixfox** [comment]
> Should this ticket be closed now that #2996 has been merged? Or is there still an issue?

No because not all age restricted videos can be bypassed so this issue is not entirely fixed.

**Comment by pukkandan** [comment]
tbh, It is doubtful this can ever can be fully fixed unless setting up an [account proxy server](https://github.com/zerodytrash/Simple-YouTube-Age-Restriction-Bypass/tree/main/account-proxy) is an acceptable option

**Comment by unixfox** [comment]
> tbh, It is doubtful this can ever can be fully fixed unless setting up an [account proxy server](https://github.com/zerodytrash/Simple-YouTube-Age-Restriction-Bypass/tree/main/account-proxy?rgh-link-date=2022-04-30T20%3A07%3A49Z) is an acceptable option

We have been talking about integrating that into invidious so it may come one day.

**Comment by unixfox** [comment]
Note: encrypted age-restricted videos don't work, see: https://github.com/iv-org/invidious/issues/3245

Example video IDs:
- https://www.youtube.com/watch?v=mQvteoFiMlg
- https://kworb.net/youtube/topvideos_restricted.html ([archive](https://web.archive.org/web/20220806131653/https://kworb.net/youtube/topvideos_restricted.html))

**Comment by SamantazFox** [comment]
Non-working video ID, reported to me by e-mail: `A2W2EBMX8DY`

**Comment by UlyssesZh** [comment]
For me, I still have issues watching age restricted videos. Loading them are significantly more slowly (which greatly influences watching) than loading other videos. I can observe this for all videos listed in [this comment](https://github.com/iv-org/invidious/issues/2189#issuecomment-1082318444).

**Comment by unixfox** [comment]
> For me, I still have issues watching age restricted videos. Loading them are significantly more slowly (which greatly influences watching) than loading other videos. I can observe this for all videos listed in [this comment](https://github.com/iv-org/invidious/issues/2189#issuecomment-1082318444).

Known issue, it's tracked in https://github.com/iv-org/invidious/issues/3245

**Comment by unixfox** [comment]
Closing as there is no way to circumvent anymore. Logging into a google account will do but it's tracked in https://github.com/iv-org/invidious/issues/4977

**Comment by unixfox** [comment]
https://github.com/TeamNewPipe/NewPipe/issues/11382#issuecomment-2628989399

**Comment by kkarhan** [comment]
> Can someone post a couple of age-restricted videos here? (this is for testing purposes though)

I've got another one if need be:
- `https://www.youtube.com/watch?v=SkXiVYlRTGQ`

I sincerely hope this gets fixed somehow.

**Comment by neurodiverseEsoteric** [comment]
https://www.youtube.com/watch?v=KgmoMO66uPg

**Comment by Fijxu** [comment]
@unixfox Do you think this is still relevant? The only fix is using Google accounts since there is no other way to get around the age restrictions with non google accounts. :/
